File: WorkcraftCore/src/main/java/org/workcraft/gui/graph/tools/HitTester.java
package org.workcraft.gui.graph.tools;

import static org.workcraft.dependencymanager.advanced.core.GlobalCache.*;

import java.awt.geom.Point2D;

import org.workcraft.dom.Node;
//import org.workcraft.dom.visual.HitMan;
import org.workcraft.dom.visual.Touchable;
import org.workcraft.dom.visual.VisualModel;
import org.workcraft.dom.visual.VisualNode;
import org.workcraft.util.Function;
import org.workcraft.util.Maybe;

import pcollections.PCollection;

public interface HitTester<N> {
	Maybe<N> hitTest(Point2D.Double point);
	PCollection<N> boxHitTest(Point2D.Double boxStart, Point2D.Double boxEnd);
	
	public class Util {
		public static <N> Function <Point2D.Double, Maybe<N>> asPointHitTester (final HitTester<N> hitTester) {
			return new Function<Point2D.Double, Maybe<N>>() {
				@Override
				public Maybe<N> apply(Point2D.Double argument) {
					return hitTester.hitTest(argument);
				}
			};
		}
		
		
	    /*		public static HitTester<VisualNode> reflectiveHitTestForSelection(final VisualModel model, final Function<? super Node, ? extends Maybe<? extends Touchable>> tp) {
			return new HitTester<VisualNode>() {
				@Override
				public Maybe<VisualNode> hitTest(Point2D.Double point) {
					return HitMan.hitTestForSelection(tp, point, (VisualNode)model.getRoot(), VisualNode.class);
				}

				@Override
				public PCollection<VisualNode> boxHitTest(Point2D.Double boxStart, Point2D.Double boxEnd) {
					return (PCollection<VisualNode>)(PCollection<?>)HitMan.boxHitTest(tp, eval(model.getRoot().children()), boxStart, boxEnd);
				}
			};
			}*/
	}
}

File: ScalaGraphEditorUtil/src/main/scala/org/workcraft/scala/grapheditor/tools/HitTester.scala
package org.workcraft.scala.grapheditor.tools
import org.workcraft.dom.visual.HitMan
import org.workcraft.dependencymanager.advanced.core.{Expression => JExpression}
import org.workcraft.dependencymanager.advanced.core.GlobalCache._
import org.workcraft.scala.Util._
import org.workcraft.scala.Expressions.Expression
import scala.collection.JavaConversions._
import java.awt.geom.Point2D
import org.workcraft.util.Maybe
import pcollections.TreePVector
import pcollections.PCollection
import org.workcraft.graphics.Touchable

trait HitTester[N] {
  def hitTest(point : Point2D.Double) : Option[N]
  def boxHitTest(boxStart : Point2D.Double, boxEnd : Point2D.Double) : List[N]
}

object HitTester {
  def create[T](nodes: Expression[_ <: Iterable[T]], touchable: T => Expression[Touchable]): HitTester[T] = {
    val transformedTouchableProvider = eval[T, Touchable](((_ : Expression[Touchable]).jexpr)compose touchable)
    new HitMan.Flat[T](eval(nodes.jexpr).toList, transformedTouchableProvider(_)).getHitTester
  }
}
